### PR TITLE
Unsigned ints no wrap

### PR DESCRIPTION
**Pull request by Shaikh-Ubaid**

### CODE REVIEW

**Comment by Shaikh-Ubaid** [comment]
This is a Work in progress (WIP). The tests might not pass.

**Comment by Shaikh-Ubaid** [comment]
```
$ cat examples/expr2.py    
from lpython import u16, bitnot_u16, i32

u: u16 = u16(255)
u = bitnot_u16(u)
print(u)
print(~i32(u))
$ lpython examples/expr2.py
65280
255
$ python examples/expr2.py 
65280
-65281
```

It seems `bitnot()` and `~` could not be successively used.

**Comment by Shaikh-Ubaid** [comment]
That [issue](https://github.com/lcompilers/lpython/pull/2176#issuecomment-1639549284) is fixed. The following is a concern:

```bash
$ cat examples/expr2.py 
from lpython import i16, u64, bitnot_u64

w: u64 = u64(123457889998923)
x: u64 = bitnot_u64(w)
print(w, i16(w), ~i16(w))
print(x, i16(x), ~i16(x))
$ python examples/expr2.py 
123457889998923 123457889998923 -123457889998924
18446620615819552692 18446620615819552692 -18446620615819552693
$ lpython examples/expr2.py
123457889998923 -26549 26548
18446620615819552692 26548 -26549
```

It seems `u8`, `u16`, `u32` and `u64` need to be some functions that could convert values appropriately in CPython.

**Comment by Shaikh-Ubaid** [comment]
If the casts in CPython become `no-op`, the output value of CPython is not matching the output of LPython (Example: https://github.com/lcompilers/lpython/pull/2176#issuecomment-1640472819).

**Comment by Shaikh-Ubaid** [comment]
For unsigned unary minus, shall we throw error in LPython similar to unsigned bit not?

**Comment by certik** [comment]
The contract is that if LPython compiles and runs this without errors, then CPython produces the same value. The design of all i8-i64, u8-u64 is such that they are just no-op in CPython and they always produce exactly the same number as in LPython, since LPython only allows such casts that don't change the value. If the value is out of range, we give an error message. So this design should work.

**Comment by Shaikh-Ubaid** [comment]
>only allows such casts that don't change the value. If the value is out of range, we give an error message.

Got it. Thank you so much for the clarification!

**Comment by certik** [comment]
> For unsigned unary minus, shall we throw error in LPython similar to unsigned bit not?

Given that -0 = 0, the unary minus should be allowed for unsigned, but obviously it will only work for 0, for all other values it must give a runtime error.

**Comment by Shaikh-Ubaid** [comment]
The CI might temporarily pass now. As a next step, we need to support 
- unsigned unary minus and uncomment the specific tests (and possibly update if the values are not in appropriate range)
- add error tests for `~` and `-`.

**Comment by Shaikh-Ubaid** [comment]
Further steps could be to support compile time overflow checking and later runtime error checking for which we have a dedicated issue here https://github.com/lcompilers/lpython/issues/2183.

File: src/runtime/lpython/lpython.py
Comment (certik) [inline src/runtime/lpython/lpython.py:717]: Why not just calling it bitnot? And make the function work for all u8 - u64.

File: src/lpython/semantics/python_ast_to_asr.cpp
Comment (certik) [inline src/lpython/semantics/python_ast_to_asr.cpp:7478]: You cannot use `or` in C++, it fails with MSVC:
```suggestion
                if (args.size() < 1 || args.size() > 1) {
```

Comment (certik) [reply src/lpython/semantics/python_ast_to_asr.cpp:7478]: Also why not just:
```suggestion
                if (args.size() != 1) {
```

**Comment by certik** [comment]
Otherwise it looks good. After you fix the above, please do not merge it, I have to manually check and update our internal codes if needed before we can merge it.

File: src/lpython/semantics/python_ast_to_asr.cpp
Comment (Shaikh-Ubaid) [reply src/lpython/semantics/python_ast_to_asr.cpp:7478]: Got it. Thank you for the guidance. I will update it.

File: src/runtime/lpython/lpython.py
Comment (certik) [reply src/runtime/lpython/lpython.py:717]: You are right, I didn't realize that. CPython puts quite a lot of limitations in what we can do, but I think it is worthy to just use the built-in `int` and `float` and build everything on top.

Let's use what you did then.

**Review by certik — APPROVED** [review]
I think this change is good. I'll mark it as draft, so that we don't accidentally merge it. I need to see if anything broke first.

**Comment by certik** [comment]
Everything works for us, so I am merging this PR.

Thanks @Shaikh-Ubaid !

**Pull request merged**
